src/cli/cli.hpp

```diff
@@ -531,10 +533,31 @@ class Interpreter : public OutputImplementer, public Output
     static void HandleTimer(Timer &aTimer);
     void        HandleTimer(void);
 
-    const otCliCommand *mUserCommands;
-    uint8_t             mUserCommandsLength;
-    void               *mUserCommandsContext;
-    bool                mCommandIsPending;
+    class otCliCommandList : public LinkedListEntry<otCliCommandList>
+    {
+        friend class LinkedListEntry<otCliCommandList>;
+
+    public:
+        void Set(const otCliCommand *aCmdList, uint8_t aLength, void *aContext)
+        {
+            mCommands       = aCmdList;
+            mCommandsLength = aLength;
+            mContext        = aContext;
+        }
+
+        const otCliCommand *mCommands;
+        uint8_t             mCommandsLength;
+        void               *mContext;
+
+    private:
+        otCliCommandList *mNext;
+    };
+
+    otCliCommandList *AllocateCommandList(void) { return mCommandListPool.Allocate(); }
+
+    LinkedList<otCliCommandList>                   mUserCommands;
+    Pool<otCliCommandList, kMaxUserCommandEntries> mCommandListPool;
+    bool                                           mCommandIsPending;
```

Comment:
We can simply define a  array here (no need to use linked-list and/or pool):

```c++
    struct UserCommandsEntry
    {
        const otCliCommand *mCommands;
        uint8_t             mLength;
        void               *mContext;
    };

    UserCommandsEntry mUserCommands[kMaxUserCommandEntries];
```

With 
```c++
void Interpreter::SetUserCommands(const otCliCommand *aCommands, uint8_t aLength, void *aContext)
{
    for (UserCommandsEntry &entry: mUserCommands)
    {
        if (entry.mCommands == nullptr)
        {
            entry.mCommands = aCommands;
            entry.mLength = aLength;
            entry.mContext = aContext;
            break;
        }
    }
}
```

And 
```c++
otError Interpreter::ProcessUserCommands(Arg aArgs[])
{
    otError error = OT_ERROR_INVALID_COMMAND;

    for (const UserCommandsEntry &entry: mUserCommands)
    {
        for (uint8_t i = 0; i < entry.mLength; i++)
        {
            if (aArgs[0] == entry.mCommands[i].mName)
            {
                char *args[kMaxArgs];

                Arg::CopyArgsToStringArray(aArgs, args);
                error = entry.mCommands[i].mCommand(entry.mContext, Arg::GetArgsLength(aArgs) - 1, args + 1);
                break;
            }
        }
    }

    return error;
}
```